File: src/main/java/NewBloc.java
import com.google.common.base.CaseFormat;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.JBColor;
import helper.BlocConfig;
import helper.BlocTaoData;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.io.*;


public class NewBloc extends AnAction {
    private Project project;
    private String psiPath;
    private BlocTaoData data;

    /**
     * Overall popup entity
     */
    private JDialog jDialog;
    private JTextField nameTextField;
    private ButtonGroup templateGroup;
    /**
     * Checkbox
     * Use folder：default true
     * Use prefix：default false
     */
    private JCheckBox folderBox, prefixBox;

    @Override
    public void actionPerformed(AnActionEvent event) {
        project = event.getProject();
        psiPath = event.getData(PlatformDataKeys.PSI_ELEMENT).toString();
        psiPath = psiPath.substring(psiPath.indexOf(":") + 1);
        initData();
        initView();
    }

    private void initData() {
        data = BlocTaoData.getInstance();
        jDialog = new JDialog(new JFrame(), "Bloc Template Code Produce");
    }

    private void initView() {
        //Set function button
        Container container = jDialog.getContentPane();
        container.setLayout(new BoxLayout(container, BoxLayout.Y_AXIS));

        //Set the main mode style
        //deal default value
        setMode(container);

        //Generate module name and OK cancel button
        setModuleAndConfirm(container);

        //Choose a pop-up style
        setJDialog();
    }

    /**
     * generate  file
     */
    private void save() {
        if (nameTextField.getText() == null || "".equals(nameTextField.getText().trim())) {
            Messages.showInfoMessage(project, "Please input the module name", "Info");
            return;
        }
        dispose();
        //Create a file
        createFile();
        //Refresh project
        project.getBaseDir().refresh(false, true);
    }

    /**
     * Set the overall pop-up style
     */
    private void setJDialog() {
        //The focus is on the current pop-up window,
        // and the focus will not shift even if you click on other areas
        jDialog.setModal(true);
        //Set padding
        ((JPanel) jDialog.getContentPane()).setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));
        jDialog.setSize(430, 350);
        jDialog.setLocationRelativeTo(null);
        jDialog.setVisible(true);
    }

    /**
     * Main module
     */
    private void setMode(Container container) {
        //Two rows and two columns
        JPanel template = new JPanel();
        template.setLayout(new GridLayout(1, 2));
        //Set the main module style：mode, function
        template.setBorder(BorderFactory.createTitledBorder("Select Mode"));
        //default: high setting
        JRadioButton defaultBtn = new JRadioButton(BlocConfig.modeDefault, data.defaultMode == 0);
        defaultBtn.setActionCommand(BlocConfig.modeDefault);
        setPadding(defaultBtn, 5, 10);
        JRadioButton highBtn = new JRadioButton(BlocConfig.modeHigh, data.defaultMode == 1);
        setPadding(highBtn, 5, 10);
        highBtn.setActionCommand(BlocConfig.modeHigh);

        template.add(defaultBtn);
        template.add(highBtn);
        templateGroup = new ButtonGroup();
        templateGroup.add(defaultBtn);
        templateGroup.add(highBtn);

        container.add(template);
        setDivision(container);
    }

    /**
     * Generate file
     */
    private void setCodeFile(Container container) {
        //Select build file
        JPanel file = new JPanel();
        file.setLayout(new GridLayout(2, 2));
        file.setBorder(BorderFactory.createTitledBorder("Select Function"));

        container.add(file);
        setDivision(container);
    }


    /**
     * Generate file name and button
     */
    private void setModuleAndConfirm(Container container) {
        JPanel nameField = new JPanel();
        nameField.setLayout(new FlowLayout());
        nameField.setBorder(BorderFactory.createTitledBorder("Module Name"));
        nameTextField = new JTextField(28);
        nameTextField.addKeyListener(keyListener);
        nameField.add(nameTextField);
        container.add(nameField);

        JPanel menu = new JPanel();
        menu.setLayout(new FlowLayout());

        //Set bottom spacing
        setDivision(container);

        //OK cancel button
        JButton cancel = new JButton("Cancel");
        cancel.setForeground(JBColor.RED);
        cancel.addActionListener(actionListener);

        JButton ok = new JButton("OK");
        ok.setForeground(JBColor.GREEN);
        ok.addActionListener(actionListener);
        menu.add(cancel);
        menu.add(ok);
        container.add(menu);
    }

    private void createFile() {
        String type = templateGroup.getSelection().getActionCommand();
        //deal default value
        if (BlocConfig.modeDefault.equals(type)) {
            data.defaultMode = 0;
        } else if (BlocConfig.modeHigh.equals(type)) {
            data.defaultMode = 1;
        }


        String name = upperCase(nameTextField.getText());
        String prefix = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, name);
        String folder =  "/" + prefix;
        String prefixName =  prefix + "_";

        switch (type) {
            case BlocConfig.modeDefault:
                generateDefault(folder, prefixName);
                break;
            case BlocConfig.modeHigh:
                generateHigh(folder, prefixName);
                break;
        }
    }

    private void generateDefault(String folder, String prefixName) {
        String path = psiPath + folder;
        String cubitPath = path+"/cubit";
        String viewPath = path+"/view";
        generateFile("cubit/view/page.dart", viewPath, prefixName + data.viewName.toLowerCase() + ".dart");
        generateFile("cubit/view/view.dart", viewPath, prefixName + data.viewFileName.toLowerCase() + ".dart");
        generateFile("cubit/cubit/state.dart", cubitPath, prefixName + "state" + ".dart");
        generateFile("cubit/cubit/cubit.dart", cubitPath, prefixName + data.cubitName.toLowerCase() + ".dart");
    }

    private void generateHigh(String folder, String prefixName) {
        String path = psiPath + folder;
        String blocPath = path+"/bloc";
        String viewPath = path+"/view";
        generateFile("bloc/view/view.dart", viewPath, prefixName + data.viewFileName.toLowerCase() + ".dart");
        generateFile("bloc/view/page.dart", viewPath, prefixName + data.viewName.toLowerCase() + ".dart");
        generateFile("bloc/bloc/bloc.dart", blocPath, prefixName + data.blocName.toLowerCase() + ".dart");
        generateFile("bloc/bloc/event.dart", blocPath, prefixName + data.eventName.toLowerCase() + ".dart");
        generateFile("bloc/bloc/state.dart", blocPath, prefixName + "state" + ".dart");
    }


    private void generateFile(String inputFileName, String filePath, String outFileName) {
        //content deal
        String content = dealContent(inputFileName, outFileName);

        //Write file
        try {
            File folder = new File(filePath);
            // if file doesnt exists, then create it
            if (!folder.exists()) {
                folder.mkdirs();
            }
            File file = new File(filePath + "/" + outFileName);
            if (!file.exists()) {
                file.createNewFile();
            }

            FileWriter fw = new FileWriter(file.getAbsoluteFile());
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(content);
            bw.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    //content need deal
    private String dealContent(String inputFileName, String outFileName) {
        //name baseFolder
        String baseFolder = "/templates/";
        String type = templateGroup.getSelection().getActionCommand();

        //read file
        String content = "";
        try {
            InputStream in = this.getClass().getResourceAsStream(baseFolder + inputFileName);
            content = new String(readStream(in));
        } catch (Exception e) {
        }

        String prefixName = "";
        //Adding a prefix requires modifying the imported class name
            prefixName = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, upperCase(nameTextField.getText())) + "_";
        //replace Cubit
        if (type.equals(BlocConfig.modeDefault)) {
            content = content.replaceAll("\\$nameCubit", "\\$name" + data.cubitName);
            content = content.replace("view.dart", prefixName + "view" + ".dart");
            content = content.replace("cubit.dart", prefixName + "cubit" + ".dart");
            content = content.replace("state.dart", prefixName + "state" + ".dart");
        }


        //replace Bloc
        if (type.equals(BlocConfig.modeHigh)) {
            content = content.replaceAll("\\$nameBloc", "\\$name" + data.blocName);
            content = content.replaceAll("\\$nameEvent", "\\$name" + data.eventName);
            content = content.replaceAll("InitEvent", "Init" + data.eventName);
            content = content.replaceAll("event.dart", prefixName + data.eventName.toLowerCase() + ".dart");
            content = content.replaceAll("event\\)", data.eventName.toLowerCase() + "\\)");
            content = content.replaceAll("event is", data.eventName.toLowerCase() + " is");
            content = content.replace("state.dart", prefixName + "state" + ".dart");
            content = content.replace("blocs.dart", prefixName + "bloc" + ".dart");
            content = content.replace("view.dart", prefixName + "view" + ".dart");

        }
        //replace Event
        if (outFileName.contains(data.eventName.toLowerCase())) {
            content = content.replaceAll("Event", data.eventName);
        }
        //replace view
        if (outFileName.contains(data.viewFileName.toLowerCase())) {
            content = content.replace("'bloc.dart'", "'" + prefixName + data.blocName.toLowerCase() + ".dart" + "'");
            content = content.replace("'cubit.dart'", "'" + prefixName + data.cubitName.toLowerCase() + ".dart" + "'");
            content = content.replace("final bloc", "final " + data.blocName.toLowerCase());
            content = content.replace("final cubit", "final " + data.cubitName.toLowerCase());
            content = content.replace("=> bloc", "=> " + data.blocName.toLowerCase());
            content = content.replace("=> cubit", "=> " + data.cubitName.toLowerCase());
            content = content.replaceAll("nameCubit", "name" + data.cubitName);
            content = content.replaceAll("nameBloc", "name" + data.blocName);
            content = content.replaceAll("'event.dart'", "'" + prefixName + data.eventName.toLowerCase() + ".dart'");
            content = content.replaceAll("Event", data.eventName);
            content = content.replace("'state.dart'", "'" + prefixName + "state.dart'");
            content = content.replace("Page", data.viewName);
        }

        content = content.replaceAll("\\$name", upperCase(nameTextField.getText()));

        return content;
    }

    private byte[] readStream(InputStream inStream) throws Exception {
        ByteArrayOutputStream outSteam = new ByteArrayOutputStream();
        try {
            byte[] buffer = new byte[1024];
            int len = -1;
            while ((len = inStream.read(buffer)) != -1) {
                outSteam.write(buffer, 0, len);
                System.out.println(new String(buffer));
            }

        } catch (IOException e) {
        } finally {
            outSteam.close();
            inStream.close();
        }
        return outSteam.toByteArray();
    }


    private final KeyListener keyListener = new KeyListener() {
        @Override
        public void keyTyped(KeyEvent e) {
        }

        @Override
        public void keyPressed(KeyEvent e) {
            if (e.getKeyCode() == KeyEvent.VK_ENTER) save();
            if (e.getKeyCode() == KeyEvent.VK_ESCAPE) dispose();
        }

        @Override
        public void keyReleased(KeyEvent e) {
        }
    };

    private final ActionListener actionListener = new ActionListener() {
        @Override
        public void actionPerformed(ActionEvent e) {
            if (e.getActionCommand().equals("Cancel")) {
                dispose();
            } else {
                save();
            }
        }
    };

    private void setPadding(JRadioButton btn, int top, int bottom) {
        btn.setBorder(BorderFactory.createEmptyBorder(top, 10, bottom, 0));
    }

    private void setMargin(JCheckBox btn, int top, int bottom) {
        btn.setBorder(BorderFactory.createEmptyBorder(top, 10, bottom, 0));
    }

    private void setDivision(Container container) {
        //Separate the spacing between modules
        JPanel margin = new JPanel();
        container.add(margin);
    }

    private void dispose() {
        jDialog.dispose();
    }

    private String upperCase(String str) {
        return str.substring(0, 1).toUpperCase() + str.substring(1);
    }
}

File: src/main/kotlin/com/flutter_code_generator/flutter_code_generator/NewBloc.kt
package com.flutter_code_generator.flutter_code_generator

import com.flutter_code_generator.flutter_code_generator.helper.BlocConfig
import com.flutter_code_generator.flutter_code_generator.helper.BlocTaoData
import com.google.common.base.CaseFormat
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.ui.JBColor
import java.awt.Container
import java.awt.FlowLayout
import java.awt.GridLayout
import java.awt.event.ActionListener
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.io.*
import java.util.*
import javax.swing.*

class NewBloc : AnAction() {
    private var project: Project? = null
    private var psiPath: String? = null
    private lateinit var data: BlocTaoData;

    /**
     * Overall popup entity
     */
    private var jDialog: JDialog? = null
    private var nameTextField: JTextField? = null
    private var templateGroup: ButtonGroup? = null

    /**
     * Checkbox
     * Use folder：default true
     * Use prefix：default false
     */
    private val folderBox: JCheckBox? = null
    private val prefixBox: JCheckBox? = null

    override fun actionPerformed(event: AnActionEvent) {
        project = event.project
        psiPath = event.getData(PlatformDataKeys.PSI_ELEMENT).toString()
        psiPath = psiPath!!.substring(psiPath!!.indexOf(":") + 1)
        initData()
        initView()
    }

    private fun initData() {
        data = BlocTaoData.instance
        jDialog = JDialog(JFrame(), "Bloc Template Code Produce")
    }

    private fun initView() {
        //Set function button
        val container = jDialog!!.contentPane
        container.layout = BoxLayout(container, BoxLayout.Y_AXIS)

        //Set the main mode style
        //deal default value
        setMode(container)

        //Generate module name and OK cancel button
        setModuleAndConfirm(container)

        //Choose a pop-up style
        setJDialog()
    }

    /**
     * generate  file
     */
    private fun save() {
        if (nameTextField!!.text == null || nameTextField!!.text.trim { it <= ' ' }.isEmpty()) {
            Messages.showInfoMessage(project, "Please input the module name", "Info")
            return
        }
        dispose()
        //Create a file
        createFile()
        //Refresh project
        project!!.baseDir.refresh(false, true)
    }

    /**
     * Set the overall pop-up style
     */
    private fun setJDialog() {
        //The focus is on the current pop-up window,
        // and the focus will not shift even if you click on other areas
        jDialog!!.isModal = true
        //Set padding
        (jDialog!!.contentPane as JPanel).border = BorderFactory.createEmptyBorder(10, 10, 10, 10)
        jDialog!!.setSize(430, 350)
        jDialog!!.setLocationRelativeTo(null)
        jDialog!!.isVisible = true
    }

    /**
     * Main module
     */
    private fun setMode(container: Container) {
        //Two rows and two columns
        val template = JPanel()
        template.layout = GridLayout(1, 2)
        //Set the main module style：mode, function
        template.border = BorderFactory.createTitledBorder("Select Mode")
        //default: high setting
        val defaultBtn: JRadioButton = JRadioButton(BlocConfig.modeDefault, data.defaultMode === 0)
        defaultBtn.actionCommand = BlocConfig.modeDefault
        setPadding(defaultBtn, 5, 10)
        val highBtn: JRadioButton = JRadioButton(BlocConfig.modeHigh, data.defaultMode === 1)
        setPadding(highBtn, 5, 10)
        highBtn.actionCommand = BlocConfig.modeHigh

        template.add(defaultBtn)
        template.add(highBtn)
        templateGroup = ButtonGroup()
        templateGroup!!.add(defaultBtn)
        templateGroup!!.add(highBtn)

        container.add(template)
        setDivision(container)
    }

    /**
     * Generate file
     */
    private fun setCodeFile(container: Container) {
        //Select build file
        val file = JPanel()
        file.layout = GridLayout(2, 2)
        file.border = BorderFactory.createTitledBorder("Select Function")

        container.add(file)
        setDivision(container)
    }


    /**
     * Generate file name and button
     */
    private fun setModuleAndConfirm(container: Container) {
        val nameField = JPanel()
        nameField.layout = FlowLayout()
        nameField.border = BorderFactory.createTitledBorder("Module Name")
        nameTextField = JTextField(28)
        nameTextField!!.addKeyListener(keyListener)
        nameField.add(nameTextField)
        container.add(nameField)

        val menu = JPanel()
        menu.layout = FlowLayout()

        //Set bottom spacing
        setDivision(container)

        //OK cancel button
        val cancel = JButton("Cancel")
        cancel.foreground = JBColor.RED
        cancel.addActionListener(actionListener)

        val ok = JButton("OK")
        ok.foreground = JBColor.GREEN
        ok.addActionListener(actionListener)
        menu.add(cancel)
        menu.add(ok)
        container.add(menu)
    }

    private fun createFile() {
        val type = templateGroup!!.selection.actionCommand
        //deal default value
        if (BlocConfig.modeDefault.equals(type)) {
            data.defaultMode = 0
        } else if (BlocConfig.modeHigh.equals(type)) {
            data.defaultMode = 1
        }


        val name = upperCase(nameTextField!!.text)
        val prefix = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, name)
        val folder = "/$prefix"
        val prefixName = prefix + "_"

        when (type) {
            BlocConfig.modeDefault -> generateDefault(folder, prefixName)
            BlocConfig.modeHigh -> generateHigh(folder, prefixName)
        }
    }

    private fun generateDefault(folder: String, prefixName: String) {
        val path = psiPath + folder
        val cubitPath = "$path/cubit"
        val viewPath = "$path/view"
        generateFile("cubit/view/page.dart", viewPath, (prefixName + data.viewName?.toLowerCase()).toString() + ".dart")
        generateFile(
            "cubit/view/view.dart",
            viewPath,
            (prefixName + data.viewFileName?.toLowerCase()).toString() + ".dart"
        )
        generateFile("cubit/cubit/state.dart", cubitPath, prefixName + "state" + ".dart")
        generateFile(
            "cubit/cubit/cubit.dart",
            cubitPath,
            (prefixName + data.cubitName?.toLowerCase()).toString() + ".dart"
        )
    }

    private fun generateHigh(folder: String, prefixName: String) {
        val path = psiPath + folder
        val blocPath = "$path/bloc"
        val viewPath = "$path/view"
        generateFile(
            "bloc/view/view.dart",
            viewPath,
            (prefixName + data.viewFileName?.toLowerCase()).toString() + ".dart"
        )
        generateFile("bloc/view/page.dart", viewPath, (prefixName + data.viewName?.toLowerCase()).toString() + ".dart")
        generateFile("bloc/bloc/bloc.dart", blocPath, (prefixName + data.blocName?.toLowerCase()).toString() + ".dart")
        generateFile("bloc/bloc/event.dart", blocPath, (prefixName + data.eventName?.toLowerCase()).toString() + ".dart")
        generateFile("bloc/bloc/state.dart", blocPath, prefixName + "state" + ".dart")
    }


    private fun generateFile(inputFileName: String, filePath: String, outFileName: String) {
        //content deal
        val content = dealContent(inputFileName, outFileName)

        //Write file
        try {
            val folder = File(filePath)
            // if file doesnt exists, then create it
            if (!folder.exists()) {
                folder.mkdirs()
            }
            val file = File("$filePath/$outFileName")
            if (!file.exists()) {
                file.createNewFile()
            }

            val fw = FileWriter(file.absoluteFile)
            val bw = BufferedWriter(fw)
            bw.write(content)
            bw.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    //content need deal
    private fun dealContent(inputFileName: String, outFileName: String): String {
        //name baseFolder
        val baseFolder = "/templates/"
        val type = templateGroup!!.selection.actionCommand

        //read file
        var content = ""
        try {
            val `in` = this.javaClass.getResourceAsStream(baseFolder + inputFileName)
            content = String(readStream(`in`))
        } catch (e: Exception) {
        }

        var prefixName = ""
        //Adding a prefix requires modifying the imported class name
        prefixName = CaseFormat.LOWER_CAMEL.to(
            CaseFormat.LOWER_UNDERSCORE, upperCase(
                nameTextField!!.text
            )
        ) + "_"
        //replace Cubit
        if (type == BlocConfig.modeDefault) {
            content = content.replace("\\\$nameCubit".toRegex(), "\\\$name" + data.cubitName)
            content = content.replace("view.dart", prefixName + "view" + ".dart")
            content = content.replace("cubit.dart", prefixName + "cubit" + ".dart")
            content = content.replace("state.dart", prefixName + "state" + ".dart")
        }


        //replace Bloc
        if (type == BlocConfig.modeHigh) {
            content = content.replace("\\\$nameBloc".toRegex(), "\\\$name" + data.blocName)
            content = content.replace("\\\$nameEvent".toRegex(), "\\\$name" + data.eventName)
            content = content.replace("InitEvent".toRegex(), "Init" + data.eventName)
            content = content.replace(
                "event.dart".toRegex(),
                (prefixName + data.eventName?.toLowerCase()).toString() + ".dart"
            )
            content = content.replace("event\\)".toRegex(), data.eventName?.toLowerCase() + "\\)")
            content = content.replace("event is".toRegex(), data.eventName?.toLowerCase() + " is")
            content = content.replace("state.dart", prefixName + "state" + ".dart")
            content = content.replace("blocs.dart", prefixName + "bloc" + ".dart")
            content = content.replace("view.dart", prefixName + "view" + ".dart")
        }
        //replace Event
        if (outFileName.contains(data.eventName!!.toLowerCase())) {
            content = content.replace("Event".toRegex(), data.eventName!!)
        }
        //replace view
        if (outFileName.contains(data.viewFileName!!.toLowerCase())) {
            content = content.replace(
                "'bloc.dart'",
                ("'" + prefixName + data.blocName?.toLowerCase()).toString() + ".dart" + "'"
            )
            content = content.replace(
                "'cubit.dart'",
                ("'" + prefixName + data.cubitName?.toLowerCase()).toString() + ".dart" + "'"
            )
            content = content.replace("final bloc", "final " + data.blocName?.toLowerCase())
            content = content.replace("final cubit", "final " + data.cubitName?.toLowerCase())
            content = content.replace("=> bloc", "=> " + data.blocName?.toLowerCase())
            content = content.replace("=> cubit", "=> " + data.cubitName?.toLowerCase())
            content = content.replace("nameCubit".toRegex(), "name" + data.cubitName)
            content = content.replace("nameBloc".toRegex(), "name" + data.blocName)
            content = content.replace(
                "'event.dart'".toRegex(),
                ("'" + prefixName + data.eventName?.toLowerCase()).toString() + ".dart'"
            )
            content = content.replace("Event".toRegex(), data.eventName!!)
            content = content.replace("'state.dart'", "'" + prefixName + "state.dart'")
            content = content.replace("Page", data.viewName!!)
        }

        content = content.replace("\\\$name".toRegex(), upperCase(nameTextField!!.text))

        return content
    }

    @Throws(Exception::class)
    private fun readStream(inStream: InputStream): ByteArray {
        val outSteam = ByteArrayOutputStream()
        try {
            val buffer = ByteArray(1024)
            var len = -1
            while ((inStream.read(buffer).also { len = it }) != -1) {
                outSteam.write(buffer, 0, len)
                println(String(buffer))
            }
        } catch (e: IOException) {
        } finally {
            outSteam.close()
            inStream.close()
        }
        return outSteam.toByteArray()
    }


    private val keyListener: KeyListener = object : KeyListener {
        override fun keyTyped(e: KeyEvent) {
        }

        override fun keyPressed(e: KeyEvent) {
            if (e.keyCode == KeyEvent.VK_ENTER) save()
            if (e.keyCode == KeyEvent.VK_ESCAPE) dispose()
        }

        override fun keyReleased(e: KeyEvent) {
        }
    }

    private val actionListener = ActionListener { e ->
        if (e.actionCommand == "Cancel") {
            dispose()
        } else {
            save()
        }
    }

    private fun setPadding(btn: JRadioButton, top: Int, bottom: Int) {
        btn.border = BorderFactory.createEmptyBorder(top, 10, bottom, 0)
    }

    private fun setMargin(btn: JCheckBox, top: Int, bottom: Int) {
        btn.border = BorderFactory.createEmptyBorder(top, 10, bottom, 0)
    }

    private fun setDivision(container: Container) {
        //Separate the spacing between modules
        val margin = JPanel()
        container.add(margin)
    }

    private fun dispose() {
        jDialog!!.dispose()
    }

    private fun upperCase(str: String): String {
        return str.substring(0, 1).uppercase(Locale.getDefault()) + str.substring(1)
    }
}